### ISSUE TITLE
Suspicious stack-overflow message that points to a valid stack range

### ISSUE THREAD

**Issue by quasilyte** [google/sanitizers#1533]
I've spent a few days trying to solve this mystery on my own, but I have to give up now and ask whether it might be a bug/false positive. Sorry if this message is too wordy, I tried to separate the important and complimentary part.

----

While running asan on some code (more details below), I got this report:

```
==168780==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7f2aa3cc6ca0 at pc 0x7f2b038b0eb5 bp 0x7f2aa3cc6bf0 sp 0x7f2aa3cc6398
WRITE of size 48 at 0x7f2aa3cc6ca0 thread T0

0x7f2aa3cc6ca0 is located 8387744 bytes inside of 8392704-byte region [0x7f2aa34c7000,0x7f2aa3cc8000)
allocated by thread T0 here:
    #0 0x7f2b03952cc7 in __interceptor_valloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10ecc7)
```

From what I see here, it looks like the allocated stack should have enough space for another 48 bytes.

* Stack offset of 8387744 is less than 8392704
* Since stack is growing downwards, we have ~8387744 bytes available
* It's also not an underflow, there are 4960 bytes above that location (the stack space that is in use)

I verified that the stack pointer looks legit and it is in fact the allocation that was performed by the application.

I do know that asan doesn't fully support makecontext API, but there are some hacks that made it possible to mitigate some false positives along the way. This is the first time I'm getting this kind of error though, with a stack overflow that looks like it's not a stack overflow.

I tried various forms of attributes to stop asan from instrumenting the problematic code path, but it doesn't look like it's helping.

I'll share more info of the code in question in a moment, but please consider the information above as a separate question that is worth answering alone. Maybe I'm just misreading the asan report and there really is something bad going on there.

Another question would be whether these annotations are intended to work recursively. I guess they don't and that would be a problem for this case (read below to get more context).

----

I'm working on a https://github.com/VKCOM/kphp compiler. Our runtime includes coroutines that are implemented with getcontext+setcontext+swapcontext. To avoid some false positives we used this API:

```
__sanitizer_finish_switch_fiber
__sanitizer_start_switch_fiber
```

Then I tried to handle another case (can be found at https://github.com/VKCOM/kphp/pull/518).

The signal handler is called on the script timeout. It's usually intercepted during the switched context (setcontext), running on a dynamically allocated stack. Then we break out of that handler using siglongjmp to a previously saved buffer somewhere inside that coroutine context (same dynamic stack). And this is where things get bad. Any attempts to allocate a stack space result in asan error, telling that it's a stack overflow. As we can see on the error messages themselves, there is a plenty of stack space in there; I verified that manually as well, by inspecting the actual SP values and comparing them to the stack space allocated for the coroutine.

It reproduces on various compiler versions; I use `g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0` locally, for instance.

There is also some `__asan_handle_no_return` warning I can't fully understand:

```
==168780==WARNING: ASan doesn't fully support makecontext/swapcontext functions and may produce false positives in some cases!
==168780==WARNING: ASan is ignoring requested __asan_handle_no_return: stack top: 0x7fffeb2ad000; bottom 0x631000023000; size: 0x1cefeb28a000 (31816768069632)
False positive error reports may follow
For details see https://github.com/google/sanitizers/issues/189
```

Let me know if I can share any more useful info.

**Comment by eugenis** [comment]
This is stack-buffer-overflow, not stack-overflow. Meaning read or write
out of bounds of some local variable on stack. If you are sure that this
particular access is within bounds and within lifetime of whatever location
it targets, then the common cause of false positives would be moving the
stack pointer up (i.e. taking a function frame or several off the stack,
like longjmp would) without telling asan about it. Sounds like your
"breaking out" of the handler is exactly that. If it's not done with
longjmp-in-libc (the one that ASan intercepts), you'd need to call
__asan_unpoison on the relevant portion of the stack.

Switching between coroutines would generally be fine with ASan. It's
undeclared returns from functions (longjmp, exceptions, etc) that asan has
problems with because we rely on each function to clear the stack shadow
after itself.

**Comment by quasilyte** [comment]
`__asan_unpoison_memory_region` does solve the issue.

Thank you for the suggestion. :+1:
